### ISSUE TITLE
Failure parsing pattern match

### ISSUE THREAD

**Issue by z5h** [elm-lang/elm-compiler#1484]
The following does not compile (parse):

```Elm
type A = A
type B = B A

f : B -> A
f b =
    case b of
        any A -> A
        _ -> A
```

**Comment by JoeyEremondi** [comment]
This is supposed to be an error. You need to pattern match on a
Constructor, and constructors all start with capital letters.

What are you trying to do?

**Comment by z5h** [comment]
@JoeyEremondi I claim that if I care about the parameter and not the constructor, then I should be able to match in a way that satisfies that. Why should that not be the case?

What I'm trying to do is beyond the scope of an SSCCE.

**Comment by jvoigtlaender** [comment]
@z5h, this is just not how pattern matching works. In Elm or Haskell or other languages with an ML style type system. 

Further discussion is almost certainly better suited to the mailing list or Slack. 

Because you are not reporting a compiler error. You want a language feature that isn't intended.

**Comment by mgold** [comment]
I see where OP is coming from. If I can match `Just 5` with `Just any`, which not with `any 5`? The problem is that if I match a `Result String Int` with `any anything`, what's the type of `anything`? Being able to match any tag's parameters clashes with the fundamental semantics of union types, which as Janis alludes have been around (under different names) since the 70s.

This sounds like a case of the [X Y Problem](http://meta.stackexchange.com/questions/66377/what-is-the-xy-problem). That is, if you follow up on the mailing list or Slack, tell us why you'd want to be able to do this, what real problem do you want to solve?

**Comment by z5h** [comment]
Thanks all.
@mgold I'm familiar with X Y problems, have my actual problem solved, but was looking to refactor my code. It seems my deviation from "fundamental semantics of union types" doesn't cause an ambiguity whereas yours does. If the compiler could distinguish between these cases I'd agree with @jvoigtlaender when he said "You want a language feature that isn't intended.". I'm curious to understand more about that and will do so via Slack or the mailing list. Cheers.
